File: samples/gdds_client.py
import uuid
import grpc
from gdds_streaming_pb2 import SubscriptionRequest
from gdds_streaming_pb2_grpc import GddsStreamServiceStub

def main():
    server_address = "{SERVER_ADDRESS}" # Replace with actual address
    api_key = "valid-api-key-123" # Replace with your API key
    token_file = "connection_token.txt"
    connection_token = None

    # Load or generate connection token
    try:
        with open(token_file, "r") as f:
            connection_token = f.read().strip()
            print(f"Using saved connection token: {connection_token}")
    except FileNotFoundError:
        connection_token = str(uuid.uuid4())
        with open(token_file, "w") as f:
            f.write(connection_token)
        print(f"Created new connection token: {connection_token}")

    try:
        with grpc.secure_channel(server_address, grpc.ssl_channel_credentials()) as channel:
            stub = GddsStreamServiceStub(channel)
            metadata = [("x-api-key", api_key)]
            request = SubscriptionRequest(connectionToken=connection_token)
            print(f"Connected with token: {connection_token}")
            for message in stub.Subscribe(request, metadata=metadata):
                print(f"Vehicle: {message.veh}, Location: ({message.lat}, {message.lon}), Time: {message.tm}, Server Time: {message.svrTm}")
    except grpc.RpcError as e:
        if e.code() == grpc.StatusCode.ALREADY_EXISTS:
            print(f"Duplicate token {connection_token}. Please generate a new one.")
        else:
            print(f"Error: {e.details()}")

if __name__ == "__main__":
    main()

File: samples/GddsClient.cs
using Grpc.Core;
using Grpc.Net.Client;
using System;
using System.IO;
using System.Threading.Tasks;

namespace GddsClient
{
    class Program
    {
        static async Task Main(string[] args)
        {
            string serverAddress = "{SERVER_ADDRESS}"; // Replace with actual address
            string apiKey = "valid-api-key-123"; // Replace with your API key
            string tokenFile = "connection_token.txt";
            string connectionToken;

            // Load or generate connection token
            if (File.Exists(tokenFile))
            {
                connectionToken = File.ReadAllText(tokenFile).Trim();
                Console.WriteLine($"Using saved connection token: {connectionToken}");
            }
            else
            {
                connectionToken = Guid.NewGuid().ToString();
                File.WriteAllText(tokenFile, connectionToken);
                Console.WriteLine($"Created new connection token: {connectionToken}");
            }

            try
            {
                using var channel = GrpcChannel.ForAddress($"https://{serverAddress}");
                var client = new GddsStreaming.GddsStreamService.GddsStreamServiceClient(channel);
                var metadata = new Metadata { { "x-api-key", apiKey } };
                var request = new SubscriptionRequest { ConnectionToken = connectionToken };

                using var call = client.Subscribe(request, metadata);
                Console.WriteLine($"Connected with token: {connectionToken}");
                await foreach (var message in call.ResponseStream.ReadAllAsync())
                {
                    Console.WriteLine($"Vehicle: {message.Veh}, Location: ({message.Lat}, {message.Lon}), Time: {message.Tm}, Server Time: {message.SvrTm}");
                }
            }
            catch (RpcException ex) when (ex.StatusCode == StatusCode.AlreadyExists)
            {
                Console.WriteLine($"Duplicate token {connectionToken}. Please generate a new one.");
            }
            catch (RpcException ex)
            {
                Console.WriteLine($"Error: {ex.Status.Detail}");
            }
        }
    }
}
